turbinia/output_writers.py

```diff
@@ -0,0 +1,170 @@
+
+import errno
+import logging
+import os
+import re
+import shutil
+import time
+
+from turbinia import config
+from turbinia import TurbiniaException
+
+from google.cloud import storage
+
+log = logging.getLogger('turbinia')
+
+
+def GetOutputWriters(result):
+  """Get a list of output writers.
+
+  Args:
+    result: A TurbiniaTaskResult object
+
+  Returns:
+    A list of OutputWriter objects.
+  """
+  epoch = str(int(time.time()))
+  log.info(u'%s %s %s' % (epoch, str(result.task_id), result.task_name))
+  unique_dir = u'{0:s}-{1:s}-{2:s}'.format(
+      epoch, str(result.task_id), result.task_name)
+
+  writers = [LocalOutputWriter(base_output_dir=result.base_output_dir,
+                               unique_dir=unique_dir)]
+  config.LoadConfig()
+  if config.GCS_OUTPUT_PATH:
+    writer = GCSOutputWriter(
+        unique_dir=unique_dir, gcs_path=config.GCS_OUTPUT_PATH)
+    writers.append(writer)
+  return writers
+
+
+class OutputWriter(object):
+  """Base class.
+
+  By default this will write the files the Evidence objects point to along with
+  any other files expclicitly written with write().
+
+  Attributes:
+    base_output_dir (string): The base path for output
+    name (string): Name of this output writer
+    unique_dir (string): A psuedo-unique string to be used in paths.
+  """
+
+  def __init__(self, unique_dir=None):
+    """Initialization for OutputWriter."""
+    self.unique_dir = unique_dir
+    self.create_output_dir()
+
+  def create_output_dir(self):
+    """Creates a unique output path for this task and creates directories.
+
+    Needs to be run at runtime so that the task creates the directory locally.
+
+    Returns:
+      A local output path string.
+
+    Raises:
+      TurbiniaException: If there are failures creating the directory.
+    """
+    raise NotImplementedError
+
+  def write(self, file_):
+    """Writes output file.
+
+    Args:
+      file_: A string path to a file.
+
+    Returns:
+      Bool indicating success
+    """
+    raise NotImplementedError
+
+
+class LocalOutputWriter(OutputWriter):
+  """Class for writing to local filesystem output.
+
+  Attributes:
+    output_dir: The full path for output.
+  """
+
+  def __init__(self, base_output_dir=None, *args, **kwargs):
+    self.base_output_dir = base_output_dir
+    self.output_dir = None
+    super(LocalOutputWriter, self).__init__(*args, **kwargs)
+    self.name = u'LocalWriter'
+
+  def create_output_dir(self):
+    self.output_dir = os.path.join(self.base_output_dir, self.unique_dir)
+    if not os.path.exists(self.output_dir):
+      try:
+        log.info(u'Creating new directory {0:s}'.format(self.output_dir))
+        os.makedirs(self.output_dir)
+      except OSError as e:
+        if e.errno == errno.EACCESS:
+          msg = u'Permission error ({0:s})'.format(str(e))
+        else:
+          msg = str(e)
+        raise TurbiniaException(msg)
+
+    return self.output_dir
+
+  def write(self, file_):
```

Comment:
I would call it file_path instead of file_ to be more descriptive of what it is.